### PR TITLE
Ruko's Player Quality Points Thing (TM)

### PR DESCRIPTION
**Pull request by ike709**
DOWNSTREAM NOTICE: This requires the latest version of rust_g (#2268).

This entire concept is solely devised by Headmin Ruko, I'm just the hired help.

TODO:

- [x] Mirror the config changes to Sage config once it's finalized
- [ ] So, so much testing
- [ ] Inevitable tweaks

## Changelog
:cl: Designed by Ruko, coded by ike709
add: Notes/bans are now quantified via a negative point value, rather than severity. Playtime as a living character gives a positive point value, effectively allowing you to improve your track record simply by playing.
/:cl:

### CODE REVIEW

**Comment by zeskorion** [comment]
We communist china now boys

**Comment by Bastian0930** [comment]
but W H Y change this?

ruko has no ping but what's the reasoning?

**Comment by Rukofamicom** [comment]
To standardize bans on a point based system by replacing the severity of notes with a point value instead. Players lose points when noted and gain them back by playing rounds without being noted. Players with more than x points after gaining new ones will be recommended (but not forced) a ban length based on that total. Points and bans will be given following a guide so that there won't be nearly as much discrepancy between different admins for the same offenses. Offenses will be worth different amounts of points within a range, so there will still be _some_, but not nearly as much. 

This will also help to highlight particularly active players for not being as problematic as they may seem - A player that breaks a rule one in ten rounds but only plays one round a week will appear to have a pretty sparse record. A player doing the same offense one in ten rounds that plays ten rounds a day will be labeled as a much worse player despite having the same overall impact on the game as the first. 

TL;DR
It's a more sophisticated way to track players who cause issues so we can gauge how much of a problem they really are. Ban lengths will be much more consistent and vary based on a player's record of good **and** bad conduct instead of just their bad conduct within any given real-time period without regard for how much they may positively contribute to the game in that same period. Permanent bans will be handed out in a more uniform fashion instead of seemingly on a whim.

**Comment by PowerfulBacon** [comment]
Why base the regaining points on how long you are playing, having is based on how many offences you have would punish players that consistently get negative points (like 13 hour player with 3 bans) would then be punished harsher than a 1000 hour player with 1 ban.

**Comment by TotallyNotChief** [comment]
muh good boy points

the question is, if i don't break any rules and if i get a lot of GBPs, can i then exchange them for chicken tendies?

**Comment by Archanial** [comment]
We cs go now

**Comment by triplezeta** [comment]
wouldn't Paradise's karma system work better as a means to regain your social credit than just playing

**Comment by AsciiSquid** [comment]
Is reading the dates and times of things really so hard as to justify an overhaul of the note system? If you really don't trust your staff to discern note volume with respect to playtime just set expiry dates for notes so they don't burden a players record months after the fact.

This change seems to be more out of a laziness in enforcing standards than anything sophisticated.

**Comment by Epsilos** [comment]
wouldn't this just encourage one admin with a grudge to collect a bunch of small 1 point infractions to throw the book at someone they hate when anyone without that grudge would only take notice when something actually worthy of a medium-high note happened

**Comment by ghost** [comment]
my social credit score...

**Comment by ethan10000llahd** [comment]
I already had a bad social credit score... ohgod no I don't want to go to brazil over low social cred-

**Comment by Rukofamicom** [comment]
> Is reading the dates and times of things really so hard

No, it isn't, but determining how much a player has played in between offenses is currently not possible at all. We can only gauge offenses against total playtime and against real-time that has passed.

> This change seems to be more out of a laziness

Is... that bad? Removing some degree of human error from the equation and making things easier on the admins in a given system is kind of the goal of programming any administrative tool isn't it?

File: SQL/database_changelog.txt [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
 
 In any query remember to add a prefix to the table names if you use one.
 
 -----------------------------------------------------
 
+Version 5.8 7 August 2020, by ike709
+Adds a new `playerqualitypoints` column in the `player` table that allows for quantifying bad behavior vs playtime.
+
+ALTER TABLE SS13_player
+`playerqualitypoints` double(4,2) ZEROFILL NOT NULL DEFAULT 0;
```
Comment (Crossedfall) [inline SQL/database_changelog.txt]:
```suggestion
ALTER TABLE SS13_player
ADD COLUMN `playerqualitypoints` double(4,2) ZEROFILL NOT NULL DEFAULT 0;
```

File: config/config.txt [hunk 2]
```diff
@@ -538,3 +538,38 @@ VOTE_AUTOTRANSFER_ENABLED
 VOTE_AUTOTRANSFER_INITIAL 54000
 ## Time (in deciseconds) between subsequent transfer votes. Default: 30 minutes
 VOTE_AUTOTRANSFER_INTERVAL 18000
+
+#### Player Quality Points Settings
```
Comment (Crossedfall) [inline config/config.txt:542]: I wonder if this should be in its own config. 🤔

File: SQL/database_changelog.txt [hunk 3]
```diff
@@ -1,15 +1,23 @@
...
 
 In any query remember to add a prefix to the table names if you use one.
 
 -----------------------------------------------------
 
+Version 5.8 7 August 2020, by ike709
+Adds a new `playerqualitypoints` column in the `player` table that allows for quantifying bad behavior vs playtime.
+
+ALTER TABLE SS13_player
+ADD COLUMN `playerqualitypoints` double(4,2) ZEROFILL NOT NULL DEFAULT 0;
```
Comment (Crossedfall) [inline SQL/database_changelog.txt]:
```suggestion
ALTER TABLE SS13_player
ADD COLUMN `playerqualitypoints` double(4,2) NOT NULL DEFAULT 0;
```

File: SQL/beestation_schema.sql [hunk 4]
```diff
@@ -308,6 +308,7 @@ CREATE TABLE IF NOT EXISTS `SS13_player` (
   `flags` smallint(5) unsigned NOT NULL DEFAULT '0',
   `antag_tokens` tinyint(4) unsigned DEFAULT '0',
   `metacoins` int(10) unsigned NOT NULL DEFAULT '0',
+  `playerqualitypoints` double(4,2) ZEROFILL NOT NULL DEFAULT 0,
```
Comment (Crossedfall) [inline SQL/beestation_schema.sql]:
```suggestion
  `playerqualitypoints` double(4,2) NOT NULL DEFAULT 0,
```

**Review by Crossedfall — COMMENTED** [review]
Gave some bad advice here, assumed we were only dealing with positive numbers.

**Comment by ike709** [comment]
This survives basic testing on staging.

**Comment by utilbird** [comment]
**what if more social credits increased your antag rep**

**Comment by MrLostman** [comment]
This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

**Comment by MCterra10** [comment]
Yes lostman, let your anger out
This is based

**Comment by utilbird** [comment]
> This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

Don't mind him, he's New Zealand

**Comment by TheFakeElon** [comment]
> This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

Don't mind him, he's New Zealand

**Comment by TheFakeElon** [comment]
> This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

Don't mind him, he's New Zealand

**Comment by TheFakeElon** [comment]
> This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

Don't mind him, he's New Zealand

**Comment by TheFakeElon** [comment]
> This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

Don't mind him, he's New Zealand

**Comment by TheFakeElon** [comment]
> This is why you shouldn't let Americans manage your code base. If you need help I'm sure AUStation could offer you some help for a price.

Don't mind him, he's New Zealand

**Comment by MCterra10** [comment]
SS13 is rapidly becoming a communist dystopia. Soon servers will share the social credit scores of their players, and players will be discriminated based on their arbitrary number oh god I can see it already

**Comment by MCterra10** [comment]
You say n word, you lose credit
You defib Captain, you get credit
Social credits as a system for in-game actions when

**Comment by Archanial** [comment]
> 
> 
> You say n word, you lose credit
> You defib Captain, you get credit
> Social credits as a system for in-game actions when

society smh

File: code/modules/jobs/job_exp.dm [hunk 5]
```diff
@@ -297,3 +298,78 @@ GLOBAL_PROTECT(exp_to_update)
...
+		list("pqp" = pqp, "ckey" = ckey)
+	)
+	query_set_pqp.warn_execute()
+	qdel(query_set_pqp)
+
+/client/proc/process_endround_playerqualitypoints()
+	if(!CONFIG_GET(flag/pqp_tracking))
+		return
+	if(GLOB.naughty_keys.len && key in GLOB.naughty_keys)
+		return
```
Comment (Crossedfall) [inline code/modules/jobs/job_exp.dm]: Travis failure

> code/modules/jobs/job_exp.dm, line 334, column 4:
> warning: ambiguous `&&` on left side of an `in`
> - add parentheses to fix: `a && (b in c)`
> - add parentheses to disambiguate: `(a && b) in c`

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Failed linting

File: code/modules/jobs/job_exp.dm [hunk 6]
```diff
@@ -297,3 +298,78 @@ GLOBAL_PROTECT(exp_to_update)
...
+		"UPDATE [format_table_name("player")] SET playerqualitypoints = :pqp WHERE ckey = :ckey",
+		list("pqp" = pqp, "ckey" = ckey)
+	)
+	query_set_pqp.warn_execute()
+	qdel(query_set_pqp)
+
+/client/proc/process_endround_playerqualitypoints()
+	if(!CONFIG_GET(flag/pqp_tracking))
+		return
+	if(GLOB.naughty_keys.len && key in GLOB.naughty_keys)
```
Comment (ike709) [inline code/modules/jobs/job_exp.dm]:
```suggestion
	if(GLOB.naughty_keys.len && (key in GLOB.naughty_keys))
```

**Comment by ike709** [comment]
Passes now.

**Comment by Archanial** [comment]
This PR scares me.

**Comment by ZombiesVsPlants** [comment]
>leaves brig as warden
>-5000 social credit

**Comment by ike709** [comment]
Just going to let this go stale. It can be reopened when the necessary policy changes happen.
